File: app/src/main/java/vedam/subkuch/base/BaseFragment.java
package vedam.subkuch.base;


import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.drawable.BitmapDrawable;
import android.location.Address;
import android.os.Build;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.PopupWindow;
import android.widget.TextView;

import androidx.annotation.AnimRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.android.volley.AuthFailureError;
import com.android.volley.NetworkError;
import com.android.volley.NetworkResponse;
import com.android.volley.ParseError;
import com.android.volley.Response;
import com.android.volley.TimeoutError;
import com.android.volley.VolleyError;
import com.crashlytics.android.Crashlytics;
import com.google.android.material.snackbar.Snackbar;
import com.google.gson.Gson;

import java.util.ArrayList;

import vedam.subkuch.R;
import vedam.subkuch.helpers.Constants;
import vedam.subkuch.interfaces.OnFragmentInteractionListener;
import vedam.subkuch.interfaces.ScreenChangeListener;
import vedam.subkuch.network.NetworkConstants;
import vedam.subkuch.network.models.ErrorResponse;
import vedam.subkuch.network.models.Image;
import vedam.subkuch.ui.matrimonial.preference.ItemAdapter;
import vedam.subkuch.uicomponent.SlideShowDialogFragment;
import vedam.subkuch.utils.LogUtils;
import vedam.subkuch.utils.UiUtil;

import static vedam.subkuch.base.BaseActivity.TAG;


/**
 * Created by nansari on 6/17/2016.
 */
public abstract class BaseFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {

    public Context context;
    private OnFragmentInteractionListener mListener;
    private ScreenChangeListener screenChangeListener;
    private SwipeRefreshLayout swipeRefreshLayout;
    protected PopupWindow mPopupWindow;

    protected Response.ErrorListener onErrorListener = error -> {

        LogUtils.LOGD("ERROR", error.getMessage());
        if (getActivity() != null)
            onErrorReceived(error);

    };

    protected void onErrorReceived(VolleyError error) {

        if (error instanceof NetworkError) {
            UiUtil.showToast(context, getString(R.string.connectionError));
        } else if (error instanceof TimeoutError) {
            UiUtil.showToast(context, getString(R.string.timeoutError));
        } else if (error instanceof ParseError) {
            UiUtil.showToast(context, getString(R.string.err_parsing));
        } else if (error instanceof AuthFailureError || (error.networkResponse != null &&
                error.networkResponse.statusCode == NetworkConstants.CODE_UNAUTHORIZED)) {
            logout();
        } else {
            parseAndShowError(error);
        }
        UiUtil.cancelProgressDialog();
    }

    void logout() {
        if (getGlobalFragmentInteractionListener() != null)
            getGlobalFragmentInteractionListener().logout();
    }
    protected void parseAndShowError(VolleyError error) {

        NetworkResponse networkResponse = error.networkResponse;
        if (networkResponse != null && networkResponse.data != null) {
            String response = new String(networkResponse.data);
            LogUtils.LOGE(TAG, "response error:" + response);

            try {
                ErrorResponse errorResponse = new Gson().fromJson(response, ErrorResponse.class);

                if (!TextUtils.isEmpty(errorResponse.getReturnMessage()))
                    UiUtil.showToast(context, errorResponse.getReturnMessage());
                else if (!TextUtils.isEmpty(errorResponse.getMessage()))
                    UiUtil.showToast(context, errorResponse.getMessage());
                else
                    UiUtil.showToast(context, getString(R.string.err_occurred));
            } catch (Exception exception) {
                Crashlytics.logException(exception);
                exception.printStackTrace();
                UiUtil.showToast(context, getString(R.string.err_occurred));
            }
        } else {
            UiUtil.showToast(context, getString(R.string.err_unknown));
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initSwipeViewIfPresent(view);
    }

    /**
     * The id should always be swipe_refresh
     *
     * @param view View inflated
     */
    private void initSwipeViewIfPresent(View view) {
//        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipe_refresh);
//        if (swipeRefreshLayout != null)
//            swipeRefreshLayout.setOnRefreshListener(this);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        this.context = context;

        if (context instanceof ScreenChangeListener)
            screenChangeListener = (ScreenChangeListener) context;
        else
            throw new RuntimeException(context.toString()
                    + " must implement ScreenChangeListener");

        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
        screenChangeListener = null;
        context = null;
    }

    protected void addFragment(final int containerId, Fragment fragment, String tag, boolean addToBackStack,
                               @AnimRes int enterAnim, @AnimRes int exitAnim,
                               @AnimRes int enterAnimPop, @AnimRes int exitAnimPop) {

        if (context instanceof BaseActivity) {
            ((BaseActivity) context).addFragment(containerId, fragment, tag, addToBackStack,
                    enterAnim, exitAnim, enterAnimPop, exitAnimPop);
        }
    }

    protected void addFragmentWithAnimation(final int containerId, Fragment fragment, String tag, boolean addToBackStack) {

        addFragment(containerId, fragment, tag, addToBackStack,
                R.anim.slide_in_right, R.anim.slide_out_left, R.anim.slide_in_left, R.anim.slide_out_right);
    }

    protected void replaceFragment(final int containerId, Fragment fragment, String tag, boolean addToBackStack,
                                   @AnimRes int enterAnim, @AnimRes int exitAnim,
                                   @AnimRes int enterAnimPop, @AnimRes int exitAnimPop) {

        if (context instanceof BaseActivity) {
            ((BaseActivity) context).replaceFragment(containerId, fragment, tag, addToBackStack,
                    enterAnim, exitAnim, enterAnimPop, exitAnimPop);
        }
    }

    /**
     * Call this method to set title of toolbar in the parent activity from fragment
     *
     * @param title Title to be set
     */
    protected void setTitle(String title) {
        if (context instanceof BaseActivity) ((BaseActivity) context).setTitle(title);

    }

    @Override
    public Context getContext() {
        return context;
    }

    public OnFragmentInteractionListener getGlobalFragmentInteractionListener() {
        return mListener;
    }


    public ScreenChangeListener getScreenChangeListener() {
        return screenChangeListener;
    }

    @Override
    public void onRefresh() {
        if (swipeRefreshLayout != null) swipeRefreshLayout.setRefreshing(false);
    }

    public void onBackPressed() {
        if (context instanceof BaseActivity) ((BaseActivity) context).onBackPressed();
    }

    /**
     * requests parent activity to get {@link Address} corresponding to current location
     */
    protected void getAddress(boolean shouldForce) {
        if (mListener != null) {
            mListener.requestAddress(shouldForce);
        }
    }

    //START_GALLERY
    protected void setGallery(ArrayList<Image> alImage, int selectedPosition, boolean isUrls) {

        Bundle bundle = new Bundle();
        bundle.putSerializable(Constants.EXTRA_IMAGE_ITEMS, alImage);
        bundle.putInt(Constants.EXTRA_POSITION, selectedPosition);
        bundle.putBoolean(Constants.EXTRA_IS_IMAGE_URLS, isUrls);

        FragmentTransaction ft = getFragmentManager().beginTransaction();
        SlideShowDialogFragment newFragment = SlideShowDialogFragment.newInstance(bundle);
        newFragment.show(ft, "slideshow");
    }

    public void baseshowFeedbackMessage(View view, String message) {
        try {
            Snackbar snakbar = Snackbar.make(view, message, Snackbar.LENGTH_LONG);
            TextView tv = snakbar.getView().findViewById(R.id.snackbar_text);
            tv.setTextColor(ContextCompat.getColor(getActivity(), R.color.colorPrimary));
            snakbar.getView().setBackgroundColor(ContextCompat.getColor(getActivity(), android.R.color.white));
            if (snakbar.isShown()) {
                snakbar.dismiss();
            }
            snakbar.show();
        } catch (Exception e) {
            Crashlytics.logException(e);
        }
    }

    public void showPopWindow(View view, ItemAdapter adapter) {
        int[] location = new int[2];
        view.getLocationOnScreen(location);
        View customView = LayoutInflater.from(getActivity()).inflate(R.layout.view_pop_window, null);
        RecyclerView recylcerView = customView.findViewById(R.id.recyclerView);
        recylcerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recylcerView.setAdapter(adapter);
        mPopupWindow = new PopupWindow(customView, view.getWidth(), WindowManager.LayoutParams.WRAP_CONTENT);
        mPopupWindow.setOutsideTouchable(true);
        mPopupWindow.setBackgroundDrawable(new BitmapDrawable());
        mPopupWindow.showAsDropDown(view, 0, 10);
    }

    public Integer checkPermission(String[] permission) {
        int permissionNeeded = 0;
        if (Build.VERSION.SDK_INT >= 23) {
            for (int i = 0; i < permission.length; i++) {
                int result = ContextCompat.checkSelfPermission(getActivity(), permission[i]);
                if (result != PackageManager.PERMISSION_GRANTED) {
                    permissionNeeded++;
                }
            }
        }
        return permissionNeeded;
    }

    protected void requestLocation() {
        if (getGlobalFragmentInteractionListener() != null)
            getGlobalFragmentInteractionListener().requestLocation(false);
    }
}

File: app/src/main/java/vedam/subkuch/base/BaseFragment.kt
package vedam.subkuch.base

import android.content.Context
import android.content.pm.PackageManager
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.WindowManager
import android.widget.PopupWindow
import android.widget.TextView
import androidx.annotation.AnimRes
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import vedam.subkuch.network.*
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.google.gson.Gson
import vedam.subkuch.R
import vedam.subkuch.helpers.Constants
import vedam.subkuch.interfaces.OnFragmentInteractionListener
import vedam.subkuch.interfaces.ScreenChangeListener
import vedam.subkuch.network.DataFetcher
import vedam.subkuch.network.NetworkConstants
import vedam.subkuch.network.models.ErrorResponse
import vedam.subkuch.network.models.Image
import vedam.subkuch.network.models.WithdrawalRequest
import vedam.subkuch.ui.jobs.models.AddResponse
import vedam.subkuch.ui.matrimonial.preference.ItemAdapter
import vedam.subkuch.uicomponent.SlideShowDialogFragment
import vedam.subkuch.utils.LogUtils
import vedam.subkuch.utils.UiUtil

/**
 * Created by nansari on 6/17/2016.
 */
abstract class BaseFragment : Fragment(), OnRefreshListener {
    @JvmField
    var mContext: Context? = null
    var globalFragmentInteractionListener: OnFragmentInteractionListener? = null
        private set
    var screenChangeListener: ScreenChangeListener? = null
        private set
    private val swipeRefreshLayout: SwipeRefreshLayout? = null

    @JvmField
    protected var mPopupWindow: PopupWindow? = null

    @JvmField
    protected var onErrorListener = Response.ErrorListener { error: ApiError ->
        LogUtils.LOGD("ERROR", error.message)
        if (activity != null) onErrorReceived(error)
    }

    protected open fun onErrorReceived(error: ApiError) {
        if (error is NetworkError) {
            UiUtil.showToast(mContext, getString(R.string.connectionError))
        } else if (error is TimeoutError) {
            UiUtil.showToast(mContext, getString(R.string.timeoutError))
        } else if (error is ParseError) {
            UiUtil.showToast(mContext, getString(R.string.err_parsing))
        } else if (error is AuthFailureError || error.networkResponse != null &&
            error.networkResponse.statusCode == NetworkConstants.CODE_UNAUTHORIZED
        ) {
            logout()
        } else {
            parseAndShowError(error)
        }
        UiUtil.cancelProgressDialog()
    }

    fun logout() {
        if (globalFragmentInteractionListener != null) globalFragmentInteractionListener!!.logout()
    }

    protected open fun parseAndShowError(error: ApiError) {
        val networkResponse = error.networkResponse
        if (networkResponse != null && networkResponse.data != null) {
            val response = String(networkResponse.data)
            LogUtils.LOGE(BaseActivity.TAG, "response error:$response")
            try {
                val errorResponse = Gson().fromJson(response, ErrorResponse::class.java)
                if (!TextUtils.isEmpty(errorResponse.returnMessage)) UiUtil.showToast(
                    mContext,
                    errorResponse.returnMessage
                ) else if (!TextUtils.isEmpty(errorResponse.message)) UiUtil.showToast(
                    mContext,
                    errorResponse.message
                ) else UiUtil.showToast(mContext, getString(R.string.err_occurred))
            } catch (exception: Exception) {
                FirebaseCrashlytics.getInstance().recordException(exception)
                exception.printStackTrace()
                UiUtil.showToast(mContext, getString(R.string.err_occurred))
            }
        } else {
            UiUtil.showToast(mContext, getString(R.string.err_unknown))
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initSwipeViewIfPresent(view)
    }

    /**
     * The id should always be swipe_refresh
     *
     * @param view View inflated
     */
    private fun initSwipeViewIfPresent(view: View) {
//        swipeRefreshLayout = (SwipeRefreshLayout) view.findViewById(R.id.swipe_refresh);
//        if (swipeRefreshLayout != null)
//            swipeRefreshLayout.setOnRefreshListener(this);
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        this.mContext = context
        screenChangeListener =
            if (context is ScreenChangeListener) context else throw RuntimeException(
                context.toString()
                        + " must implement ScreenChangeListener"
            )
        if (context is OnFragmentInteractionListener) {
            globalFragmentInteractionListener = context
        } else {
            throw RuntimeException(
                context.toString()
                        + " must implement OnFragmentInteractionListener"
            )
        }
    }

    override fun onDetach() {
        super.onDetach()
        globalFragmentInteractionListener = null
        screenChangeListener = null
        mContext = null
    }

    protected fun addFragment(
        containerId: Int, fragment: Fragment?, tag: String?, addToBackStack: Boolean,
        @AnimRes enterAnim: Int, @AnimRes exitAnim: Int,
        @AnimRes enterAnimPop: Int, @AnimRes exitAnimPop: Int
    ) {
        if (mContext is BaseActivity) {
            (mContext as BaseActivity).addFragment(
                containerId, fragment, tag, addToBackStack,
                enterAnim, exitAnim, enterAnimPop, exitAnimPop
            )
        }
    }

    protected fun addFragmentWithAnimation(
        containerId: Int,
        fragment: Fragment?,
        tag: String?,
        addToBackStack: Boolean
    ) {
        addFragment(
            containerId,
            fragment,
            tag,
            addToBackStack,
            R.anim.slide_in_right,
            R.anim.slide_out_left,
            R.anim.slide_in_left,
            R.anim.slide_out_right
        )
    }

    protected fun replaceFragment(
        containerId: Int, fragment: Fragment?, tag: String?, addToBackStack: Boolean,
        @AnimRes enterAnim: Int, @AnimRes exitAnim: Int,
        @AnimRes enterAnimPop: Int, @AnimRes exitAnimPop: Int
    ) {
        if (mContext is BaseActivity) {
            (mContext as BaseActivity).replaceFragment(
                containerId, fragment, tag, addToBackStack,
                enterAnim, exitAnim, enterAnimPop, exitAnimPop
            )
        }
    }

    /**
     * Call this method to set title of toolbar in the parent activity from fragment
     *
     * @param title Title to be set
     */
    protected fun setTitle(title: String?) {
        if (mContext is BaseActivity) (mContext as BaseActivity).title = title!!
    }

    override fun getContext(): Context? {
        return mContext
    }

    override fun onRefresh() {
        if (swipeRefreshLayout != null) swipeRefreshLayout.isRefreshing = false
    }

    fun onBackPressed() {
        if (mContext is BaseActivity) (mContext as BaseActivity).navigateBack()
    }

    /**
     * requests parent activity to get [Address] corresponding to current location
     */
    protected fun getAddress(shouldForce: Boolean) {
        if (globalFragmentInteractionListener != null) {
            globalFragmentInteractionListener!!.requestAddress(shouldForce)
        }
    }

    //START_GALLERY
    protected fun setGallery(alImage: ArrayList<Image?>?, selectedPosition: Int, isUrls: Boolean) {
        val bundle = Bundle()
        bundle.putSerializable(Constants.EXTRA_IMAGE_ITEMS, alImage)
        bundle.putInt(Constants.EXTRA_POSITION, selectedPosition)
        bundle.putBoolean(Constants.EXTRA_IS_IMAGE_URLS, isUrls)
        val ft = fragmentManager!!.beginTransaction()
        val newFragment = SlideShowDialogFragment.newInstance(bundle)
        newFragment.show(ft, "slideshow")
    }

    fun baseshowFeedbackMessage(view: View?, message: String?) {
        try {
            val snakbar = Snackbar.make(
                view!!, message!!, Snackbar.LENGTH_LONG
            )
            val tv = snakbar.view.findViewById<TextView>(com.google.android.material.R.id.snackbar_text)
            tv.setTextColor(ContextCompat.getColor(activity!!, R.color.colorPrimary))
            snakbar.view.setBackgroundColor(
                ContextCompat.getColor(
                    activity!!,
                    android.R.color.white
                )
            )
            if (snakbar.isShown) {
                snakbar.dismiss()
            }
            snakbar.show()
        } catch (e: Exception) {
            FirebaseCrashlytics.getInstance().recordException(e)
        }
    }

    fun showPopWindow(view: View, adapter: ItemAdapter?) {
        val location = IntArray(2)
        view.getLocationOnScreen(location)
        val customView = LayoutInflater.from(activity).inflate(R.layout.view_pop_window, null)
        val recylcerView = customView.findViewById<RecyclerView>(R.id.recyclerView)
        recylcerView.layoutManager = LinearLayoutManager(activity)
        recylcerView.adapter = adapter
        mPopupWindow = PopupWindow(customView, view.width, WindowManager.LayoutParams.WRAP_CONTENT)
        mPopupWindow!!.isOutsideTouchable = true
        mPopupWindow!!.setBackgroundDrawable(BitmapDrawable())
        mPopupWindow!!.showAsDropDown(view, 0, 10)
    }

    fun checkPermission(permission: Array<String?>): Int {
        var permissionNeeded = 0
        if (Build.VERSION.SDK_INT >= 23) {
            for (i in permission.indices) {
                val result = ContextCompat.checkSelfPermission(activity!!, permission[i]!!)
                if (result != PackageManager.PERMISSION_GRANTED) {
                    permissionNeeded++
                }
            }
        }
        return permissionNeeded
    }

    protected fun requestLocation() {
        if (globalFragmentInteractionListener != null) globalFragmentInteractionListener!!.requestLocation(
            false
        )
    }

    open fun withdraw() {
        UiUtil.showProgressDialog(mContext, getString(R.string.please_wait))
        val withdrawalRequest = WithdrawalRequest()
        withdrawalRequest.setAmount("10")
        withdrawalRequest.setVendorCode("180204")
        DataFetcher.withdraw(
            mContext, Gson().toJson(withdrawalRequest), onWithdrawalSuccessListener,
            AddResponse::class.java, onErrorListener
        )
    }

    private val onWithdrawalSuccessListener =
        Response.Listener { response: AddResponse? ->
            UiUtil.cancelProgressDialog()
            var errorMessage: String? = getString(R.string.err_occurred)
            if (activity == null)
                return@Listener
            if (response != null && !TextUtils.isEmpty(response.returnMessage)) errorMessage =
                response.returnMessage
            if (response != null && response.returnMessage == Constants.SUCCESS) {
                moneyWithdrawn()
            } else UiUtil.showToast(mContext, errorMessage!!)
        }

    open fun moneyWithdrawn() {}
}
